src/main/jbake/content/blog/2016/12/19/hawkular-metrics-compression.adoc

```diff
@@ -0,0 +1,93 @@
+
+=== Compression job
+
+A while ago, we introduced a job scheduler to the Hawkular-Metrics and it's been improved on 0.20.0 and 0.21.0 based on our requirements to do reliable repeated compression jobs. When the server launches it schedules a repeated job called `CompressionData` that is ran every 2 hours at the start of odd hours. Should you for some reason miss one execution (because the server down for example), the scheduler will immediately start the missing jobs in order with only one being active at a time (cluster wide, state is maintained in Cassandra). The scheduled job `CompressionData` will take care of our ETL process for the whole cluster (this is something we will also improve later so that one can scale the amount of compression jobs, but for now we deemed it to be enough fast).
+
+When the job is started, the job itself takes a look at the intended runtime (irrelevant of the actual running time as that could be influenced by server downtime or previous job) and starts to fetch all the metrics that were created during the last 2 even hour block. For example, assuming the job should start at 03:00, we will compress the datapoints that were sent between 00:00 and 02:00. This allows up to one hour of out-of-order writes to be still compressed. If an out-of-order writes happens after this, it is kept uncompressed in the Cassandra and merged during the read phase.
+
+We have limited the compression job to only process a single time serie at once, to reduce the memory pressure on the Cassandra node as it might have to load existing data to memory from the disk. To reduce the I/O effect you can allocate more memory to the Cassandra as that will make reading from the disk less likely and can also reduce the amount of compaction jobs that are required as already compressed data doesn't have to be written to the disk. Should any Cassandra request fail during the compression job, we will retry the operation with a linear wait period between the requests. 
+
+All the datapoints that were read are then streamed to the compression method of choice out of heap to reduce GC pressure. We place a header as the first byte in the compressed array, with 4 first bits reserved for the compression type and last 4 bits as metadata for the compression type. 
+
+=== Reading pipeline
+
+Since we store the data now potentially in two different places, `data` for uncompressed metrics and `data_compressed` for compressed metrics, we need to read from both if necessary. While usually the `data` table holds newer datapoints than `data_compressed` this is not guaranteed as we support writing out-of-order writes without any limitations. 
```

Comment:
You may explain above that the compression removes data from `data` and writes into `data_compressed`.